Lib/test/test_unparse.py

```diff
@@ -130,7 +130,32 @@ class Foo: pass
 
 class ASTTestCase(unittest.TestCase):
     def assertASTEqual(self, ast1, ast2):
-        self.assertEqual(ast.dump(ast1), ast.dump(ast2))
+        # Ensure the comparisons start at an AST node
+        self.assertIsInstance(ast1, ast.AST)
+        self.assertIsInstance(ast2, ast.AST)
+
+        # An AST comparison routine modeled after ast.dump(), but
+        # instead of string building, it traverses the two trees
+        # in lock-step.  Defining an inner function also prevents
+        # the `missing` value from polluting the "real" function
+        # signature or module namespace.
+        missing = object()
+        def traverse_compare(a, b):
+            if type(a) is not type(b):
+                self.fail("type(a) != type(b)")
+            if isinstance(a, ast.AST):
+                for field in a._fields:
+                    value1 = getattr(a, field, missing)
+                    value2 = getattr(b, field, missing)
+                    traverse_compare(value1, value2)
```

Comment:
You can avoid the call if both are missing

Reply:
True, but that adds an additonal `if` for the successful case, which IMO should be the common path.

Reply:
A function call is extremely more expensive than a conditional, specially if you compare them with `is`.

Reply:
As I thought, adding a check for `missing` ended up with a statistically insignificant slowdown, however it did lead to the realization of singleton checking, which results in a further 2% improvement.